std/algorithm/comparison.d

```diff
@@ -913,81 +913,80 @@ template equal(alias pred = "a == b")
+                return false;
             }
-        }
-        //Try a fast implementation when the ranges have comparable lengths
-        else static if (hasLength!Range1 && hasLength!Range2 && is(typeof(r1.length == r2.length)))
-        {
-            immutable len1 = r1.length;
-            immutable len2 = r2.length;
-            if (len1 != len2) return false; //Short circuit return
+            //Detect default pred and compatible dynamic array
+            else static if (is(typeof(pred) == string) && pred == "a == b" &&
+                isArray!Range1 && isArray!Range2 && is(typeof(r1 == r2)))
+            {
+                return r1 == r2;
+            }
+            // if one of the arguments is a string and the other isn't, then auto-decoding
+            // can be avoided if they have the same ElementEncodingType
+            else static if (is(typeof(pred) == string) && pred == "a == b" &&
+                isAutodecodableString!Range1 != isAutodecodableString!Range2 &&
+                is(immutable ElementEncodingType!Range1 == immutable ElementEncodingType!Range2))
+            {
+                import std.utf : byCodeUnit;
 
-            //Lengths are the same, so we need to do an actual comparison
-            //Good news is we can squeeze out a bit of performance by not checking if r2 is empty
-            for (; !r1.empty; r1.popFront(), r2.popFront())
+                static if (isAutodecodableString!Range1)
+                {
+                    return equal(r1.byCodeUnit, r2);
+                }
+                else
+                {
+                    return equal(r2.byCodeUnit, r1);
+                }
+            }
+            //Try a fast implementation when the ranges have comparable lengths
+            else static if (hasLength!Range1 && hasLength!Range2 && is(typeof(r1.length == r2.length)))
+            {
+                immutable len1 = r1.length;
+                immutable len2 = r2.length;
+                if (len1 != len2) return false; //Short circuit return
+
+                //Lengths are the same, so we need to do an actual comparison
+                //Good news is we can squeeze out a bit of performance by not checking if r2 is empty
+                for (; !r1.empty; r1.popFront(), r2.popFront())
+                {
+                    if (!binaryFun!(pred)(r1.front, r2.front)) return false;
+                }
+                return true;
+            }
+            else
             {
-                if (!binaryFun!(pred)(r1.front, r2.front)) return false;
+                //Generic case, we have to walk both ranges making sure neither is empty
+                for (; !r1.empty; r1.popFront(), r2.popFront())
+                {
+                    if (r2.empty) return false;
+                    if (!binaryFun!(pred)(r1.front, r2.front)) return false;
+                }
+                static if (!isInfinite!Range1)
+                    return r2.empty;
```

Comment:
The hunk above looks a lot of a bigger change than it is... I just merge the two overloads into one with a static if inside.